oximeter/db/tests/integration_test.rs

```diff
@@ -0,0 +1,380 @@
+        request_timeout,
+    );
+    wait_for_ping(&client1).await?;
+    wait_for_ping(&client2).await?;
+    wait_for_keepers(&deployment, (1..=3).collect()).await?;
+    let end = tokio::time::Instant::now();
+    println!("deploy setup time = {:?}", end - start);
+
+    let start = tokio::time::Instant::now();
+    client1
+        .init_test_minimal_replicated_db()
+        .await
+        .context("failed to initialize db")?;
+    let end = tokio::time::Instant::now();
+    println!("init replicated db time = {:?}", end - start);
+
+    // Ensure our database tables show up on both servers
+    let start = tokio::time::Instant::now();
+    let output1 = client1.list_replicated_tables().await?;
+    let output2 = client2.list_replicated_tables().await?;
+    let end = tokio::time::Instant::now();
+    println!("list tables time = {:?}", end - start);
+    assert_eq!(output1, output2);
+
+    let input = TestInput::default();
+
+    // Let's write some samples to our first replica and wait for them to show
+    // up on replica 2.
+    let start = tokio::time::Instant::now();
+    let samples = test_util::generate_test_samples(
+        input.n_projects,
+        input.n_instances,
+        input.n_cpus,
+        input.n_samples,
+    );
+    let end = tokio::time::Instant::now();
+    println!("generate samples time = {:?}", end - start);
+    assert_eq!(samples.len(), input.n_points());
+    let start = tokio::time::Instant::now();
+    client1.insert_samples(&samples).await.expect("failed to insert samples");
+    let end = tokio::time::Instant::now();
+    println!("insert samples time = {:?}", end - start);
+
+    // Get all the samples from the replica where the data was inserted
+    let start = tokio::time::Instant::now();
+    let oxql_res1 = client1
+        .oxql_query("get virtual_machine:cpu_busy")
+        .await
+        .expect("failed to get all samples");
+    let end = tokio::time::Instant::now();
+    println!("query samples from client1 time = {:?}", end - start);
+
+    // Ensure the samples are correct on this replica
+    assert_input_and_output(&input, &samples, &oxql_res1);
+
+    let start = tokio::time::Instant::now();
+    wait_for_num_points(&client2, samples.len())
+        .await
+        .expect("failed to get samples from client2");
+    let end = tokio::time::Instant::now();
+    println!("query samples from client2 time = {:?}", end - start);
+
+    // Add a 3rd clickhouse server and wait for it to come up
+    deployment.add_server().expect("failed to launch a 3rd clickhouse server");
+    let client3 = Client::new_with_request_timeout(
+        deployment.http_addr(3)?,
+        log,
+        request_timeout,
+    );
+    wait_for_ping(&client3).await?;
+
+    // We need to initiate copying from existing replicated tables by creating
+    // the DB and those tables on the new node.
+    client3
+        .init_test_minimal_replicated_db()
+        .await
+        .expect("failed to initialized db");
```

Comment:
It seems like we might want this to be part of the cluster deployment in some way, rather than a separate step. Clickward itself doesn't need to care about the SQL, but the _oximeter_ deployment should probably be running with the SQL files that setup the database automatically.

I've been thinking about this for a while. It's technically possible to create a `Client` and then init _both_ the replicated and single-node databases, which seems like it should not be possible. I've not thought very hard about how to prevent that, but maybe we can do that as we start to integrate Clickward.

Reply:
I'm not sure this is necessarily how we want that to work. Reconfigurator is going to take over initializing the databases and ensuring compatibility and upgrade, right?

Reply:
Yeah, I agree it's going to be reconfigurator's job. What I meant was it would be nice if the abstraction of the deployment did not require the caller to specify more than once whether it's a single-node or replicated cluster. As an example, suppose in reconfigurator, we added a `ClickHouseDeployment` type (intentionally different from the Clickward type). What I'm suggesting is something like this:

```rust
let deployment = ClickHouseDeployment::cluster(cluster_details);
```

Rather than:

```rust
let deployment = ClickHouseDeployment::cluster(cluster_details);
client.init_replicated_db(depolyment.http_addr(1));
```

That call to initialize the database should happen inside the cluster setup, so that it's not possible (or very hard) to run the single-node SQL against a cluster. You can do that today quite easily.

That's all :) I don't think there's any changes needed here, just a direction I think we might want to move towards.

Reply:
Makes sense!